finality-aleph/src/validator_network.rs

```diff
@@ -0,0 +1,41 @@
+use aleph_primitives::AuthorityId;
+use futures::future::pending;
+
+/// Network represents an interface for opening and closing connections with other Validators,
+/// and sending direct messages between them.
+///
+/// Note on Network reliability and security: it is neither assumed that the sent messages must be
+/// always delivered, nor the established connections must be secure in any way. If the Network
+/// implementation fails to deliver a message, it may assume that the send method will be called
+/// again.
+#[async_trait::async_trait]
+pub trait Network<A, D>: Send {
+    /// Add the peer to the set of connected peers.
+    fn add_connection(&mut self, peer: AuthorityId, addresses: Vec<A>);
+
+    /// Remove the peer from the set of connected peers and close the connection.
+    fn remove_connection(&mut self, peer: AuthorityId);
```

Comment:
do we assume that add[remove]_functions always succeed?  If not change return type to smth like Result<(), NetworkError>

Reply:
And add errors to reflect  such cases: peer not in the connections, peer already in the connections etc

Reply:
The main implementation will likely just use channels, so it's very hard to return meaningful errors here, unfortunately. :/

Reply:
still imho trait should at least try to make it possible even though main implementation may ignore it

Reply:
> do we assume that add[remove]_functions always succeed?
I think we might assume that, because the intention is not to have an established connection after the method returns, but to only add/remove the address to an internal hashmap
> If not change return type to smth like Result<(), NetworkError>
In this case we'd have to make these methods `async`, and we'd rather not...

Reply:
Would it need to be an async though? Feel free to resolve this (and send) conversation but I still feel like here and in case of send we can make the trait a little more open to possible fails even though the implementation may just not bother with 'fail' execution path.

Reply:
yeah, but in the end it would be more misleading to have this if we do not even plan to implement returning `Err`, so I think it's best to be honest :)